neural_speed/convert/convert_phi.py

```diff
@@ -0,0 +1,307 @@
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+# Convert Hugging Face fine-tuned gpt-neox-like models to ne format
+#
+# Usage:
+#
+#   python3 models/convert-h5-to-ne.py
+#
+# This script is similar to "convert-pt-to-ne.py"
+#
+
+import struct
+import numpy as np
+from pathlib import Path
+import argparse
+from typing import (IO, TYPE_CHECKING, Any, Callable, Dict, Iterable, List, Literal, Optional, Sequence, Tuple, TypeVar,
+                    Union)
+from transformers import AutoModelForCausalLM, AutoTokenizer
+import gguf
+
+# ref: https://github.com/openai/gpt-2/blob/master/src/encoder.py
+def bytes_to_unicode():
+    """
+    Returns list of utf-8 byte and a corresponding list of unicode strings.
+    The reversible bpe codes work on unicode strings.
+    This means you need a large # of unicode characters in your vocab if you want to avoid UNKs.
+    When you're at something like a 10B token dataset you end up needing around 5K for decent coverage.
+    This is a significant percentage of your normal, say, 32K bpe vocab.
+    To avoid that, we want lookup tables between utf-8 bytes and unicode strings.
+    And avoids mapping to whitespace/control characters the bpe code barfs on.
+    """
+    bs = list(range(ord("!"), ord("~") + 1)) + list(range(ord("¡"), ord("¬") + 1)) + list(range(ord("®"), ord("ÿ") + 1))
+    cs = bs[:]
+    n = 0
+    for b in range(2**8):
+        if b not in bs:
+            bs.append(b)
+            cs.append(2**8 + n)
+            n += 1
+    cs = [chr(n) for n in cs]
+    return dict(zip(bs, cs))
+        
+def phi_convert_gguf(model, tokenizer, dir_model, fname_out, ftype, hparams):
+    print("phi.gguf converting: ")
+    list_vars = model.state_dict()
+    n_rot = int(hparams["partial_rotary_factor"]*hparams["hidden_size"]/hparams["num_attention_heads"])
+    for name in list_vars.keys():
+        print(name, list_vars[name].shape, list_vars[name].dtype)
+
+    print(hparams)
+
+    gguf_file = fname_out + '.gguf'
+    gguf_writer = gguf.GGUFWriter(gguf_file, "phi")
+
+    gguf_writer.add_uint32('magic', 0x67676d66)
+    gguf_writer.add_uint32('version', 1)
+    gguf_writer.add_uint32('n_vocab', hparams["vocab_size"])
+    gguf_writer.add_embedding_length(hparams["hidden_size"])
+    gguf_writer.add_uint32('n_mult', 0)
+    gguf_writer.add_head_count(hparams["num_attention_heads"])
+    gguf_writer.add_head_count_kv(hparams["num_key_value_heads"])
+
+    gguf_writer.add_block_count(hparams["num_hidden_layers"])
+    gguf_writer.add_rope_dimension_count(n_rot)
+    gguf_writer.add_uint32('ftype', ftype)
+    gguf_writer.add_context_length(hparams["max_position_embeddings"])
+    gguf_writer.add_max_alibi_bias(0)
+    gguf_writer.add_uint32('clip_qkv', 0)
```

Comment:
If you decide to use GGUF as the convert format, you don't need to add these useless parameters. Please remove these unused parameters and try again.

such as clip_qkv, par_res...

Reply:
done